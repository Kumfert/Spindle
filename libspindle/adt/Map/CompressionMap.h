//
//  CompressionMap.h -- an abstraction of a compression mapping
//
//  $Id: CompressionMap.h,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  This class is meant to abstract away the details of a compression
//  mapping.  It has a fine2coarse vector and a coarse2fine matrix.
//  Given one, it will create the other.
//
//  A CompressionMap is not in a valid state until validate() is called
//  to insure that both fine2coarse and coarse2fine are consistent with 
//  each other.
//

#ifndef SPINDLE_COMPRESSION_MAP_H_
#define SPINDLE_COMPRESSION_MAP_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_MATRIX_BASE_H_
#include "spindle/MatrixBase.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class CompressionMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( CompressionMap )
    
private:
  CompressionMap();
  bool validateFine2Coarse();
  bool validateCoarse2Fine();
  void createFine2Coarse();
  void createCoarse2Fine();
  SharedArray<int>      fine2Coarse;
  SharedPtr<MatrixBase> coarse2Fine;

  enum { NEITHER, FINE2COARSE, COARSE2FINE } changedMostRecent;
  int szCoarse;
  int szFine;
public:
  CompressionMap( const int nUncompressed, const int nCompressed );
  CompressionMap( const CompressionMap& cmap );
  virtual ~CompressionMap();

  int queryNCompressed() const;
  int queryNUncompressed() const;

  const SharedArray<int>& getFine2Coarse() const { return fine2Coarse; }
  SharedArray<int>& getFine2Coarse() { changedMostRecent=FINE2COARSE; return fine2Coarse; }

  const SharedPtr<MatrixBase>& getCoarse2Fine() const { return coarse2Fine; }
  SharedPtr<MatrixBase>& getCoarse2Fine() { changedMostRecent=COARSE2FINE; return coarse2Fine; }

  virtual void validate();
  virtual void dump( FILE * fp ) const;
  virtual bool reset();
  bool resize(  const int nUncompressed, const int nCompressed );

};

inline
int CompressionMap::queryNCompressed() const {
  return szCoarse;
}

inline 
int CompressionMap::queryNUncompressed() const {
  return szFine;
}

inline
CompressionMap::CompressionMap( const int nUncompressed, const int nCompressed ) {
  incrementInstanceCount( CompressionMap::MetaData );
  reset();
  coarse2Fine.take( new MatrixBase );
  resize( nUncompressed, nCompressed );
}

inline
CompressionMap::CompressionMap( const CompressionMap& cmap ) 
  : szCoarse( cmap.szCoarse ), szFine( cmap.szFine ) {
    incrementInstanceCount( CompressionMap::MetaData );
    fine2Coarse.import( cmap.fine2Coarse );
    coarse2Fine.import( cmap.coarse2Fine.operator->() );
    changedMostRecent = cmap.changedMostRecent;
    currentState = cmap.currentState;
}

inline
CompressionMap::~CompressionMap() {
  decrementInstanceCount( CompressionMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif
