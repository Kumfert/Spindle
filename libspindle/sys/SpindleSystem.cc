//
// SpindleSystem.cc
//
// $Id: SpindleSystem.cc,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//

#include "spindle/SpindleSystem.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

static int count = 0;

static FILE* TraceFile = stderr;
static FILE* WarnFile = stderr;
static FILE* ErrorFile = stderr;

static unsigned int TraceLevel = 0;
static PersistanceRegistry * Registry = 0;
static StackTrace * Stack = 0;

static char* currentFunction;
static char* currentFile;
static int currentLine;

PersistanceRegistry& 
SpindleSystem::registry() {
  return *Registry;
}

StackTrace& 
SpindleSystem::stack() {
  return *Stack;
}

void
SpindleSystem::startSystem() {
  if (count++ == 0) {
    Registry = new PersistanceRegistry();
    Stack = new StackTrace();
  }
}

void
SpindleSystem::stopSystem() {
  if ( --count==0 ) {
    delete Registry;
    delete Stack;
  }
}



void
SpindleSystem::setWhere( char* func, char* file, int line ) {
  currentFunction = func;
  currentFile = file;
  currentLine = line;
}

void 
SpindleSystem::setTraceFile( FILE* traceFile ) {
  TraceFile = traceFile;
}

void 
SpindleSystem::setWarnFile( FILE* warnFile ) {
  WarnFile = warnFile;
}


void 
SpindleSystem::setErrorFile( FILE* errorFile ) {
  ErrorFile = errorFile;
}

void 
SpindleSystem::setTraceLevel( unsigned int traceLevel ) {
  if ( traceLevel >= 0 ) {
    TraceLevel = traceLevel;
  }
}

unsigned int 
SpindleSystem::getTraceLevel() { 
  return TraceLevel;
}

//
// Now implement the print functions for standard libraries.
//
//
#if SPINDLE_TARGET == std

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( (! arg)  && (ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(ErrorFile, "\nASSERT FAILED:  ");
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
    SpindleSystem::stack().reset();
#else
    SpindleSystem::printWhere( ErrorFile );
#endif
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg && ( WarnFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else 
    SpindleSystem::printWhere( WarnFile );
#endif
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

//
// else make a library using matlab constructs...
//
//
#elif SPINDLE_TARGET == matlab 

extern "C" {
#include "mex.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg ) {
    va_list ap;
    va_start( ap, format);
    if ( ErrorFile == 0 )  { // print via Matlab's messaging system.
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {  // assume that a file is open
      fprintf(ErrorFile, "\nASSERT FAILED:  ");
      vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( ErrorFile );
      SpindleSystem::stack().reset();
#else
      SpindleSystem::printWhere( ErrorFile );
#endif
    }
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile == 0 ) {
      int i = sprintf( buffer, "\nASSERT FAILED:  ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%ld):  ",errorCode);
    vfprintf(ErrorFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( ErrorFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile == 0 ) {
      vsprintf( buffer, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
      mexWarnMsgTxt( buffer );
  } else if ( ErrorFile != 0 ) {
    fprintf(ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile == 0 ) {
      int i = sprintf( buffer, "\nWARNING:   ");
      vsprintf( buffer+i, format, ap );
      mexWarnMsgTxt( buffer );
    } else {
      fprintf(WarnFile, "\nWARNING:  ");
      vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
      SpindleSystem::stack().dump( WarnFile );
#else 
      SpindleSystem::printWhere( WarnFile );
#endif
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile == 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf( buffer+i, format, ap );
    mexWarnMsgTxt( buffer );
  } else if ( WarnFile != 0 ) {
    fprintf(WarnFile, "\nWARNING:  ");
    vfprintf(WarnFile, format, ap);
#ifdef SPINDLE_ENABLE_STACK_TRACE
    SpindleSystem::stack().dump( WarnFile );
#else
    printWhere( ErrorFile );
#endif
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile == 0 ) {
      vsprintf( buffer, format, ap );
      mexPrintf( "%s", buffer );
    } else if ( TraceFile != 0 ) {
      vfprintf(TraceFile, format, ap);
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( TraceFile == 0 ) {
    vsprintf( buffer, format, ap );
    mexPrintf( "%s", buffer );
  } else  if ( TraceFile != 0 ) {
    vfprintf(TraceFile, format, ap);
  }
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    fprintf(fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    sprintf( buffer, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
    mexPrintf( "%s", buffer );
  }
}

#elif  SPINDLE_TARGET == petsc 

extern "C" {
#include "petsc.h"
}
static char buffer[512*8];

void 
SpindleSystem::assert_that( bool arg, const char* format, ... ) {
  if ( ! arg  && ( ErrorFile != 0 ) ) {
    va_list ap;
    va_start( ap, format);
    int i = sprintf( buffer, "\nASSERT FAILED:  ");
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    va_end(ap);
    // exit(-1);
  }
}

void 
SpindleSystem::error( unsigned long errorCode, const char* format, ... ) {
  va_list ap;
  va_start( ap, format );
  if ( ErrorFile != 0 ) {
    int i = sprintf( buffer, "\nERROR(%ld):  ", errorCode );
    vsprintf( buffer+i, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "%s", buffer );
    SpindleSystem::printWhere( ErrorFile );
    printWhere( ErrorFile );
  }
  va_end(ap);
}

merr
SpindleSystem::mError(  int line, char* file, char* func, merr errcode, char* msg ) {
  if ( ErrorFile != 0 ) {
    PetscFPrintf( PETSC_COMM_SELF, ErrorFile, "\nERROR(%d): %s in \"%s\" %s[%d]",errcode, msg, func, file, line );
  }
  return errcode;
}

void 
SpindleSystem::warn_if( bool arg, const char* format, ... ) {
  if ( arg ) {
    va_list ap;
    va_start( ap, format);
    if ( WarnFile != 0 ) {
      int i = sprintf( buffer, "\nWARNING:  ");
      vsprintf( buffer+i, format, ap );
      PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
      SpindleSystem::printWhere( WarnFile );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::warning( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  if ( WarnFile != 0 ) {
    int i = sprintf( buffer, "\nWARNING:  ");
    vsprintf(buffer+i, WarnFile, format, ap);
    PetscFPrintf( PETSC_COMM_SELF, WarnFile, "%s", buffer );
    printWhere( WarnFile );
  }
  va_end(ap);
}

void 
SpindleSystem::trace( unsigned int traceLevel, const char* format, ... ) {
  if ( traceLevel & TraceLevel ) {
    va_list ap;
    va_start( ap, format);
    //    const StackInfo& si = SpindleSystem::stack().top();
    //    if ( traceLevel != SPINDLE_TRACE_STACK ) {
    //      fprintf(TraceFile, "\nTrace(%d) %s [%d] - \"%s\"\n\t",traceLevel, si.file, si.line, si.function );
    //    }
    if ( TraceFile != 0 ) {
      vsprintf(buffer , format, ap);
      PetscFPrintf( PETSC_COMM_SELF, TraceFile, "%s", buffer );
    }
    va_end(ap);
  }
}

void 
SpindleSystem::printf( const char* format, ... ) {
  va_list ap;
  va_start( ap, format);
  vsprintf( buffer , format, ap );
  PetscPrintf( PETSC_COMM_SELF,  "%s", buffer );
  va_end(ap);
}

void 
SpindleSystem::printWhere( FILE * fp ) {
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp != 0 ) ) {
    PetscFPrintf( PETSC_COMM_SELF, fp, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
  if ( (currentFunction != 0) && (currentFile != 0 ) && (fp == 0 ) ) {
    PetscPrintf( PETSC_COMM_SELF, "\n \"%s\" in %s-%d\n",currentFunction, currentFile, currentLine );
  }
}

#else
#error Error do not recognize SPINDLE TARGET of SPINDLE_TARGET
#endif
