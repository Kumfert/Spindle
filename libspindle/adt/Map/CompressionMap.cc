//
// CompressionMap.cc
//
//  $Id: CompressionMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#include "spindle/CompressionMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

CompressionMap::CompressionMap() {
  incrementInstanceCount( CompressionMap::MetaData );
  coarse2Fine.take( new MatrixBase );
  reset();
}

bool
CompressionMap::reset() {
  currentState = EMPTY;
  return true;
}

bool 
CompressionMap:: resize( const int nUncompressed, const int nCompressed ) {
  if ( currentState == INVALID ) { return false; }
  if ( nUncompressed <= 0 ) { return false; }
  if ( nCompressed <= 0 ) { return false; }
  if ( nCompressed > nUncompressed ) { return false; }
  szCoarse = nCompressed;
  szFine = nUncompressed;
  fine2Coarse.resize( szFine );
  coarse2Fine->resize( szCoarse, szFine, szFine );
  currentState = UNKNOWN;
  return true;
}

void
CompressionMap::validate() {
  if ( currentState == VALID ) { return; }
  if ( currentState == INVALID ) { return; }
  bool fine2CoarseOkay = false;
  bool coarse2FineOkay = false;
  if ( changedMostRecent == NEITHER ) {
    // pick one
    if ( fine2CoarseOkay = validateFine2Coarse() ) {
      changedMostRecent = FINE2COARSE;
    } else if ( coarse2FineOkay = validateCoarse2Fine() ) {
      changedMostRecent = COARSE2FINE;
    } else {
      currentState = INVALID;
      return;
    }
  }
  if ( changedMostRecent == FINE2COARSE ) {
    // validate fine2coarse if neccessary
    if ( !fine2CoarseOkay ) {
      if ( !validateFine2Coarse() ) {
	currentState = INVALID;
	return;
      }
    }
    // create coarse2fine
    createCoarse2Fine();
  } else if ( changedMostRecent == COARSE2FINE ) {
    // validate coarse2Fine if neccessary
    if ( !coarse2FineOkay ) {
      if ( !validateCoarse2Fine() ) {
	currentState = INVALID;
	return;
      }
    }
    // create fine2coarse
    createFine2Coarse();
  } else {
    // error
    currentState = INVALID;
    return;
  }
  currentState = VALID;
  return;
}

bool
CompressionMap::validateFine2Coarse() {
  // If reported size and actual size differ.
  // the user must have changed the actual size
  // most recent.
  if ( szFine < fine2Coarse.size() ) { 
    currentState = INVALID;
    return false;
  }
  int max=0;
  int stop = fine2Coarse.size();
  const int * cur = fine2Coarse.begin();
  {for ( int i=0; i<stop; ++i, ++cur ) {
    if ( *cur < 0 ) {
      currentState = INVALID;
      return false;
    }
    max = ( *cur > max ) ? *cur : max ;
  }}
  szFine = fine2Coarse.size();
  szCoarse = max + 1; // 0 to n-1
  return true;
}

bool
CompressionMap::validateCoarse2Fine() {
  int nC = coarse2Fine->getColPtr().size() - 1;
  int n = coarse2Fine->getRowIdx().size();
  if ( nC > n ) { 
    currentState = INVALID;
    return false;
  }

  // const int * colPtr = coarse2Fine->getColPtr().lend(); // unused 
  const int * rowIdx = coarse2Fine->getRowIdx().lend();
  SharedArray<int> Temp( n );
  Temp.init( 0 );
  int * temp = Temp.begin();

  // check that each uncompressed vertex is represented.
  {for ( int i=0; i<n; ++i ) {
    temp[ rowIdx[i] ]++;
  }}
  {for ( int i=0; i<n; ++i ) {
    if (temp[ i ] != 1 ) {
      currentState = INVALID;
      return false;
    }
  }}
  szCoarse = nC;
  szFine = n;
  return true;
}


void 
CompressionMap::createCoarse2Fine() {
  int n = szFine;
  int nC = szCoarse;
  const int * fine2coarse = fine2Coarse.begin();

  // first find the max in fine2Coarse;
  coarse2Fine->getColPtr().resize( nC + 1);
  coarse2Fine->getColPtr().init( 0 );
  coarse2Fine->getRowIdx().resize( n );
  SharedArray<int> Temp( n );

  int * colPtr = coarse2Fine->getColPtr().begin();
  int * rowIdx = coarse2Fine->getRowIdx().begin();
  int * temp = Temp.begin();
  
  // now construct reverse map
  {for(int i=0;i<n;++i) { // count instances of each supernode
    colPtr[fine2coarse[i]+1]++;
  }}
  {for(int i=0;i<nC;++i) { // compute a cumulative sum
    colPtr[i+1]+=colPtr[i];
    temp[i] = colPtr[i];  // temp hold the next available spot in head column
  } }
  {for(int i=0;i<n;++i) {
    int j = fine2coarse[i];
    rowIdx[ temp[j]++ ] = i;
  }}
  
  // check
  // {for(i=0;i<nC;++i) {
  //   if( temp[i] != adjhead[i+1]) ERROR(SPINDLE_ERROR, "Indices do not add up!");
  // }}
  return;
}

void 
CompressionMap::createFine2Coarse() {
  int nC = szCoarse;
  //  int n = szFine; // unused 
  fine2Coarse.resize( szFine );
  fine2Coarse.init( -1 );
  const int * colPtr = coarse2Fine->getColPtr().lend();
  const int * rowIdx = coarse2Fine->getRowIdx().lend();
  int * fine2coarse = fine2Coarse.begin();
  {for (int i=0; i<nC; ++i) {
    for (int jj=colPtr[i]; jj<colPtr[i+1]; ++jj ) {
      int j = rowIdx[jj];
      fine2coarse[j]=i;
    }
  }}
}

void
CompressionMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"CompressionMap:");
    const int * p = fine2Coarse.begin(); 
    const int n = fine2Coarse.size();
    for( int i=0; i<n; i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
CompressionMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = fine2Coarse.lend();
    size_t my_size = fine2Coarse.size();
    pair< const int *, size_t> my_pair( p, my_size );
    ar << my_pair;
  } else { 
    ar << -1;
  }
}

void
CompressionMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else { 
    fine2Coarse.resize( new_size );
    int* p = fine2Coarse.begin();
    pair< int*, size_t > new_pair( p, new_size );
    
    ar >> new_pair;
    
    changedMostRecent = FINE2COARSE;
    validate();
  }
}

SPINDLE_IMPLEMENT_PERSISTANT( CompressionMap, SpindlePersistant )
