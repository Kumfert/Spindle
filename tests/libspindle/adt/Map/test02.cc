//
// src/adt/Map/drivers/tests/test01.cc -- tests Compression
//
//  $Id: test02.cc,v 1.2 2000/02/18 01:32:07 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//


#include "spindle/CompressionMap.h"

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#include <iostream.h>
#include <iomanip.h>

#define REGISTRY SpindleSystem::registry()

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  ios::sync_with_stdio();
  SpindleSystem::registry().setDumpOnDestroy( stdout );
  SpindleSystem::setErrorFile( stdout );

  int comp_array[] = { 0, 0, 1, 0, 1, 1, 2, 3, 4, 4, 3, 4 };
  CompressionMap cmap( 12, 5 );
  cmap.getFine2Coarse().import( comp_array, 12 );
  cmap.validate();

  if ( ! cmap.isValid() ) { 
    cerr << "WARNING: cannot create valid CompressionMap." << endl;
  }

  cout << "Size of CompressionMap class = " << sizeof(CompressionMap) << endl;
  
  const ClassMetaData* metaData = REGISTRY.findClass("CompressionMap");
  
  cout << "According to the registry it is = " << metaData->classSize << endl;
  cout << "This cmap object is serial #" << cmap.queryIDNumber() << endl;
  
  cout << "Dumping the Object { " << endl << "  ";
  cmap.dump( stdout );
  cout << endl << "}" << endl;
  cout << "coarse2fine map = {" << endl;
  const int * colPtr = cmap.getCoarse2Fine()->getColPtr().lend();
  const int * rowIdx = cmap.getCoarse2Fine()->getRowIdx().lend();
  const int n = cmap.getCoarse2Fine()->getColPtr().size() - 1;
  cout << "  nCoarse = " << n << endl;
  {for ( int i = 0; i<n; ++i ) { 
    cout << "  " << setw(5) << i << ":  ";
    for ( int jj=colPtr[i]; jj<colPtr[i+1]; ++jj ) { 
      int j = rowIdx[jj];
      cout << setw(5) << j << " ";
    }
    cout << endl;
  }}
  cout << "}" << endl;

}
