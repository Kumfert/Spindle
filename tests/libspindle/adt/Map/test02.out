Size of CompressionMap class = 44
According to the registry it is = 44
This cmap object is serial #0
Dumping the Object { 
  CompressionMap:0 0 1 0 1 1 2 3 4 4 3 4 
}
coarse2fine map = {
  nCoarse = 5
      0:      0     1     3 
      1:      2     4     5 
      2:      6 
      3:      7    10 
      4:      8     9    11 
}
Destroying Registry... PersistanceRegistery::dump():  3 items
             CompressionMap (44) 1/1---> SpindlePersistant
           SpindleBaseClass (8) 0/0---> 
          SpindlePersistant (12) 1/1---> SpindleBaseClass
